File: Binary Search Tree/106.Construct Binary Tree from Inorder and Postorder Traversal.py
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        def buildTreePostorder(inorder,ins,ine,
                              postorder,pos,poe,map1):
            if pos>poe or ins>ine:
                return None
            
            root=TreeNode(postorder[poe])
            inroot=map1[postorder[poe]]
            numleft=inroot-ins
            
            root.left=buildTreePostorder(inorder,ins,inroot-1,
                                        postorder,pos,pos+numleft-1,map1)
            
            root.right=buildTreePostorder(inorder,inroot+1,ine,
                                         postorder,pos+numleft,poe-1,map1)
            return root
            
            
        
        if len(inorder)!=len(postorder):
            return None
        
        map1={}
        for i in range(len(inorder)):
            map1[inorder[i]]=i
        return buildTreePostorder(inorder,0,len(inorder)-1,
                                 postorder,0,len(postorder)-1,map1)
        
            
        

File: Binary Search Tree/Java/106. Construct Binary Tree from Inorder and Postorder Traversal.java
*******************************106. Construct Binary Tree from Inorder and Postorder Traversal*****************************************

class Solution {
    public TreeNode buildTree(int[] inorder,int inStart,int inEnd, int[] postorder,int postStart,int postEnd,HashMap<Integer,Integer> map)
    {
        if(inStart>inEnd || postStart>postEnd) return null;

        TreeNode root=new TreeNode(postorder[postEnd]);
        int inRoot=map.get(root.val);
        int numLeft=inRoot-inStart;
        //post->Left,right,root
    
        root.left=buildTree(inorder,inStart,inRoot-1,postorder,postStart,postStart+numLeft-1,map);
        root.right=buildTree(inorder,inRoot+1,inEnd,postorder,postStart+numLeft,postEnd-1,map);

        return root;
    }

    public TreeNode buildTree(int[] inorder, int[] postorder) {
        HashMap<Integer,Integer> map=new HashMap<>();
        for(int i=0;i<inorder.length;i++)
        {
            map.put(inorder[i],i);
        }

        TreeNode root=buildTree(inorder,0,inorder.length-1,postorder,0,postorder.length-1,map);
        return root;

    }
}